go/vt/vtgate/planbuilder/operators/ast2op.go

```diff
@@ -115,16 +115,205 @@ func createOperatorFromUpdate(ctx *plancontext.PlanningContext, updStmt *sqlpars
+	updClone := sqlparser.CloneRefOfUpdate(updStmt)
+	updOp, err := createUpdateOperator(ctx, updStmt, vindexTable, qt, routing)
 	if err != nil {
 		return nil, err
 	}
 
+	ksMode, err := ctx.VSchema.ForeignKeyMode(vindexTable.Keyspace.Name)
+	if err != nil {
+		return nil, err
+	}
+	// Unmanaged foreign-key-mode, we don't need to do anything.
+	if ksMode != vschemapb.Keyspace_FK_MANAGED {
+		return updOp, nil
+	}
+
+	parentFKs := vindexTable.ParentFKsNeedsHandling()
+	childFks := vindexTable.ChildFKsNeedsHandling(vindexes.UpdateAction)
+	if len(childFks) == 0 && len(parentFKs) == 0 {
+		return updOp, nil
+	}
+
+	parentFKs, childFks = fkNeedsHandlingForUpdates(updStmt.Exprs, parentFKs, childFks)
+	if len(parentFKs) > 0 {
+		return nil, vterrors.VT12003()
+	}
+
+	// If there are no foreign key constraints, then we don't need to do anything.
+	if len(childFks) == 0 {
+		return updOp, nil
+	}
+
+	// If the delete statement has a limit, we don't support it yet.
+	if updStmt.Limit != nil {
+		return nil, vterrors.VT12001("foreign keys management at vitess with limit")
+	}
+
+	var fkChildren []*FkChild
+	var selectExprs []sqlparser.SelectExpr
+	for _, fk := range childFks {
+		// Any RESTRICT type foreign keys that arrive here,
+		// are cross-shard/cross-keyspace RESTRICT cases, which we don't currently support.
+		if isRestrict(fk.OnUpdate) {
+			return nil, vterrors.VT12002()
+		}
+
+		// We need to select all the parent columns for the foreign key constraint, to use in the update of the child table.
+		var cols []int
+		for _, column := range fk.ParentColumns {
+			cols = append(cols, len(selectExprs))
+			selectExprs = append(selectExprs, aeWrap(sqlparser.NewColName(column.String())))
+		}
+
+		// We only support literals in update queries, which trigger foreign key updates.
+		for _, expr := range updStmt.Exprs {
+			colIdx := fk.ParentColumns.FindColumn(expr.Name.Name)
+			if colIdx < 0 {
+				continue
+			}
+			switch expr.Expr.(type) {
+			case *sqlparser.Argument, *sqlparser.NullVal, sqlparser.BoolVal, *sqlparser.Literal:
+			default:
+				return nil, vterrors.VT12001("foreign keys management at vitess with non-literal values")
+			}
+		}
+
+		// We now construct the update query for the child table -
+		// For CASCADE type constraint, the query looks like this -
+		// 		`UPDATE <child_table> SET <child_column_updated_using_update_exprs_from_parent_update_query> WHERE <child_columns_in_fk> IN (<bind variable for the output from SELECT>)`
+		// For SET NULL type constraint, the query looks like this -
+		//		`UPDATE <child_table> SET <child_column_updated_using_update_exprs_from_parent_update_query>
+		//		WHERE <child_columns_in_fk> IN (<bind variable for the output from SELECT>)
+		//		[AND <child_columns_in_fk> NOT IN (<bind variables in the SET clause of the original update>)]`
+		// 	If all the columns of the foreign key in the parent don't actually end up getting updated, then we shouldn't be setting the child columns to NULL.
```

Comment:
I have added a lot of comments to this block of code because this isn't straightforward.